### PR TITLE
fiptool: Add --pad command line option

### PR DESCRIPTION
**Pull request by dp-arm**
Allow each image to reserve some padding space before alignment.

Use case is --align 0x4000 --pad 64 to make sure sufficient
room is available to prepend (or append) vendor-specific headers.

Helps resolve afaerber/meson-tools#3.

Change-Id: I31e59456600b72cd30cde0f653ec6523480a5823
Signed-off-by: Andreas Färber <afaerber@suse.de>
Signed-off-by: Dimitris Papastamos <dimitris.papastamos@arm.com>

### CODE REVIEW

**Comment by dp-arm** [comment]
> Thanks, but "dp-arm" is not an acceptable Signed-off-by.

All of my patches use that form.

**Comment by afaerber** [comment]
It violates contributing.md and the Linux DCO.

**Comment by masahir0y** [comment]
> Use case is --align 0x4000 --pad 64 to make sure sufficient
room is available to prepend (or append) vendor-specific headers.

From this statement, it was not clear whether `--pad` means `--pre-pad` or `--post-pad`.

If I understood the source code correctly, it looks like `--post-pad`.
It it OK if you are sure `--pre-pad` is unneeded.

Here, I mean `--pre-pad` aligns the start of extra space,
and `--post-pad` aligns the start of image bodies.

**Comment by dp-arm** [comment]
> Here, I mean --pre-pad aligns the start of extra space,
> and --post-pad aligns the start of image bodies.

Yes, I think it is --post-pad according to this definition.  I guess the reason --pre-pad is not going to be particularly useful is because the image will not run correctly if the start of the image contains extra headers (offsets will be incorrect).  There's no magic done in the TF loader to cope with this situation.

**Comment by afaerber** [comment]
Actually the Amlogic firmware does need a pre-pad, but the proprietary tool (and currently my tool) assumes a post-pad. For a --pre-pad much more code would need to be changed than just the alignment lines. Not sure if it's worth implementing that, given that a simple cat command could achieve the same effect for a fixed size header. On the other hand, by that argument, I could just cat a dd if=/dev/zero of=pad bs=1 count=N file and use unmodified --align and drop the PR. Thoughts?

**Comment by masahir0y** [comment]
> On the other hand, by that argument, I could just cat a dd if=/dev/zero of=pad bs=1 count=N file to the end of my blobs and use unmodified --align and drop the PR. Thoughts

Yeah, I was also thinking this.

Append (or prepend) headers, then put the images into FIP
 will be easier than
Make extra spaces in FIP, then embed headers

File: tools/fiptool/fiptool.c [hunk 1]
```diff
@@ -729,7 +749,8 @@ static void create_usage(void)
 	printf("fiptool create [opts] FIP_FILENAME\n");
 	printf("\n");
 	printf("Options:\n");
-	printf("  --align <value>\t\tEach image is aligned to <value> (default: 1).\n");
+	printf("  --align <bytes>\t\tEach image is aligned to <bytes> (default: 1).\n");
+	printf("  --pad <bytes>\t\t\tEach image is padded by <bytes> (default: 0).\n");
```
Comment (danh-arm) [inline tools/fiptool/fiptool.c]: minor: Could say "... padded at the end by...". Same for other cases below.

**Comment by danh-arm** [comment]
Hi @afaerber. We don't have a strong view on whether it's better to use `dd` before putting images into the FIP or to use the feature in this PR. If you say this PR satisfies your needs and you will use it, we will merge, otherwise we'll drop it.

**Comment by danh-arm** [comment]
jenkins: retest this please

**Comment by dp-arm** [comment]
@afaerber ping

**Comment by davidcunado-arm** [comment]
@afaerber  
We're still pending your feedback on this PR. Please could you confirm that this PR satisfies your needs and you will use it? 
I'll look to either close it or merge it by Wednesday 28-JUN depending on what your feedback is.

**Comment by davidcunado-arm** [comment]
Closing PR - it has not been confirmed that this change will be used. Please re-open if this PR is still required.

**Pull request closed without merging**